flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala

```diff
@@ -334,6 +335,156 @@ class Table(
+
+  /**
+    * Joins two [[Table]]s. Similar to an SQL left outer join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]] and its [[TableConfig]] must
+    * have nullCheck enables.
+    *
+    * Example:
+    *
+    * {{{
+    *   left.leftOuterJoin(right, 'a === 'b && 'c > 3).select('a, 'b, 'd)
+    * }}}
+    */
+  def leftOuterJoin(right: Table, joinPredicate: Expression): Table = {
+    join(right, joinPredicate, JoinType.LEFT_OUTER)
+  }
+
+  /**
+    * Joins two [[Table]]s. Similar to an SQL right outer join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]] and its [[TableConfig]] must
+    * have nullCheck enables.
+    *
+    * Example:
+    *
+    * {{{
+    *   left.rightOuterJoin(right, "a = b && c > 3").select('a, 'b, 'd)
+    * }}}
+    */
+  def rightOuterJoin(right: Table, joinPredicate: String): Table = {
+    join(right, joinPredicate, JoinType.RIGHT_OUTER)
+  }
+
+  /**
+    * Joins two [[Table]]s. Similar to an SQL right outer join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]] and its [[TableConfig]] must
+    * have nullCheck enables.
+    *
+    * Example:
+    *
+    * {{{
+    *   left.rightOuterJoin(right, 'a === 'b && 'c > 3).select('a, 'b, 'd)
+    * }}}
+    */
+  def rightOuterJoin(right: Table, joinPredicate: Expression): Table = {
+    join(right, joinPredicate, JoinType.RIGHT_OUTER)
+  }
+
+  /**
+    * Joins two [[Table]]s. Similar to an SQL full outer join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]] and its [[TableConfig]] must
+    * have nullCheck enables.
+    *
+    * Example:
+    *
+    * {{{
+    *   left.fullOuterJoin(right, "a = b && c > 3").select('a, 'b, 'd)
+    * }}}
+    */
+  def fullOuterJoin(right: Table, joinPredicate: String): Table = {
+    join(right, joinPredicate, JoinType.FULL_OUTER)
+  }
+
+  /**
+    * Joins two [[Table]]s. Similar to an SQL full outer join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]] and its [[TableConfig]] must
+    * have nullCheck enables.
+    *
+    * Example:
+    *
+    * {{{
+    *   left.fullOuterJoin(right, 'a === 'b && 'c > 3).select('a, 'b, 'd)
+    * }}}
+    */
+  def fullOuterJoin(right: Table, joinPredicate: Expression): Table = {
+    join(right, joinPredicate, JoinType.FULL_OUTER)
+  }
+
+  private def flinkJoinTypeToCalcite(joinType: JoinType) = joinType match {
```

Comment:
I think you can also directly use Calcite's `JoinRelType` instead of Flink's `JoinType`.